activerecord/lib/active_record/relation/where_clause.rb

```diff
@@ -107,9 +111,12 @@ def equality_node?(node)
         end
 
         def non_conflicting_binds(other)
-          conflicts = referenced_columns & other.referenced_columns
-          conflicts.map! { |node| node.name.to_s }
-          binds.reject { |attr| conflicts.include?(attr.name) }
+          # Choose all predicates that had bound params to work with them based on index
+          predicates_with_binds = predicates.select { |predicate| equality_node?(predicate) && predicate.right.is_a?(Arel::Nodes::BindParam) }
```

Comment:
This sounds like it's assuming BindParam can only appear as an RHS of an equality predicate -- otherwise the bind indexes wouldn't match up.

But is that really true, in the general case? 

Maybe I'm missing some reason we can make strict assumptions about the predicates here, but it seems to me that we'd need to do a full walk to identify the binds they're each using.